x-pack/plugins/fleet/server/services/package_policies_to_agent_permissions.ts

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { SavedObjectsClientContract } from 'kibana/server';
+
+import type { FullAgentPolicyOutputPermissions } from '../../common';
+import { getPackageInfo } from '../../server/services/epm/packages';
+
+import type { PackagePolicy } from '../types';
+
+export const DEFAULT_PERMISSIONS = {
+  cluster: ['monitor'],
+  indices: [
+    {
+      names: [
+        'logs-*',
+        'metrics-*',
+        'traces-*',
+        'synthetics-*',
+        '.logs-endpoint.diagnostic.collection-*',
+      ],
+      privileges: ['auto_configure', 'create_doc'],
+    },
+  ],
+};
+
+export async function storedPackagePoliciesToAgentPermissions(
+  soClient: SavedObjectsClientContract,
+  packagePolicies: string[] | PackagePolicy[]
+): Promise<FullAgentPolicyOutputPermissions | undefined> {
+  if (packagePolicies.length === 0) {
+    return;
+  }
+
+  // I'm not sure what permissions to return for this case, so let's return the defaults
+  if (typeof packagePolicies[0] === 'string') {
+    return { _fallback: DEFAULT_PERMISSIONS };
+  }
+
+  const permissionEntries = (packagePolicies as PackagePolicy[]).map<Promise<[string, any]>>(
+    async (packagePolicy) => {
+      if (!packagePolicy.package) {
+        return [packagePolicy.name, DEFAULT_PERMISSIONS];
+      }
+
+      const pkg = await getPackageInfo({
+        savedObjectsClient: soClient,
+        pkgName: packagePolicy.package.name,
+        pkgVersion: packagePolicy.package.version,
+      });
+
+      // FIXME is this true?
+      if (!pkg.data_streams) {
+        return [packagePolicy.name, DEFAULT_PERMISSIONS];
+      }
+
+      let dataStreamsForPermissions: DataStreamMeta[];
+
+      switch (pkg.name) {
+        case 'endpoint':
```

Comment:
If possible, we should not have any special cases in here. Instead I guess these package will need some additional configurations? Same for apm, happy to chat.

Reply:
The problem is not in the `package`, but in the `packagePolicy`. When an integration is installed, a `packagePolicy` object is created and linked to the policy. This `packagePolicy` has `inputs`, these inputs have `streams`, and these streams have a `data_stream`, which loosely corresponds with a package's `data_stream`.

My understanding is that the user can enable/disable these `inputs` (i.e. when installing the `apache` integration, a user might choose to collect logs but not metrics).

APM and Endpoint `inputs` don't have `streams` (and therefore `data_streams`), so with things as they are now I'm assuming all data_streams from the package are needed.

I think the next step would be to revisit the data associated with a `packagePolicy` and try to remove all current exceptions.